### PR TITLE
feat: addition of upgrader for application factor settings

### PR DESCRIPTION
**Pull request by leleueri**
this upgrade list all the application and if factors are enable, initialize the new factor settings data structures

  old values remains unchanged

fixes AM-1228

### CODE REVIEW

File: gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        mfaSettings.setEnroll(enrollSettings);
+        enrollSettings.setActive(true);
+        if (!mfaSettings.getEnrollment().getForceEnrollment()) {
+            enrollSettings.setType(MfaEnrollType.OPTIONAL);
+            enrollSettings.setForceEnrollment(false);
+            enrollSettings.setSkipTimeSeconds(mfaSettings.getEnrollment().getSkipTimeSeconds());
+        } else {
+            enrollSettings.setType(MfaEnrollType.REQUIRED);
+            enrollSettings.setForceEnrollment(true);
+        }
```
Comment (mstasgravitee) [inline gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java:133]: MfaEnrollType.CONDITIONAL is missing

File: gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    protected Single<Boolean> processUpgrade(String instanceOperationId, SystemTask task, String previousOperationId) {
+        return updateSystemTask(task, (SystemTaskStatus.ONGOING), previousOperationId)
+                .flatMap(updatedTask -> {
+                    if (updatedTask.getOperationId().equals(instanceOperationId)) {
+                        return migrateFactorSettings(updatedTask);
+                    } else {
+                        return Single.error(new IllegalStateException("Task " + getTaskId() + " already processed by another instance : trigger a retry"));
+                    }
+                })
+                .map(__ -> true);
```
Comment (mstasgravitee) [inline gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java:76]:
```suggestion
           ;
```

File: gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    private static void moveFactorIdsIntoFactorSettings(Application app, List<String> recoveryCodeIds) {
+        final var factorSettings = new HashSet<ApplicationFactorSettings>();
+        boolean defaultFactor = false;
+        for (var id : app.getFactors()) {
+            if (!defaultFactor) {
+                // default factor need to be the first factor from the list
+                // but recovery code factor shouldn't be a default candidate
+                defaultFactor = !recoveryCodeIds.contains(id);
+            }
+            factorSettings.add(new ApplicationFactorSettings(id, null, defaultFactor));
```
Comment (mstasgravitee) [inline gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java:172]: application factor settings has been changed a few days ago:

this is the new version. It is used inside FactorMongoSettings
@Getter
@Setter
@NoArgsConstructor
public class ApplicationFactorSettings {
    private String id;
    private String selectionRule;

    public ApplicationFactorSettings(ApplicationFactorSettings applicationFactorSettings) {
        this.id = applicationFactorSettings.id;
        this.selectionRule = applicationFactorSettings.selectionRule;
    }
}


Please use latest destination branch code.

File: gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                    logger.error("Unable to migrate factor settings for applications: {}", err.getMessage());
+                    return Single.just(false);
+                });
+    }
+
+    private static MFASettings migrateEnrollSettings(Application app) {
+        final var mfaSettings = app.getSettings().getMfa();
+        final var enrollSettings = new EnrollSettings();
+        mfaSettings.setEnroll(enrollSettings);
+        enrollSettings.setActive(true);
```
Comment (mstasgravitee) [inline gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java:125]: mfa can be turned off in previous version

File: gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            enrollSettings.setForceEnrollment(true);
+        }
+        return mfaSettings;
+    }
+
+    private static void migrateChallengeSettings(Application app) {
+        final var mfaSettings = app.getSettings().getMfa();
+        final var challengeSettings = new ChallengeSettings();
+        mfaSettings.setChallenge(challengeSettings);
+        challengeSettings.setActive(true);
```
Comment (mstasgravitee) [inline gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java:141]: mfa can be turned off in previous version

**Review by mstasgravitee — CHANGES_REQUESTED** [review]

File: gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java (hunk 1 above)
Comment (leleueri) [reply gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java:133]: In 4.2 and previous version, enrollment is either REQUIRED or OPTIONAL. When CONDITIONAL rule is defined then it apply only on challenge.

**Review by mstasgravitee — APPROVED** [review]

**Pull request merged**